hedera-node/cli-clients/src/main/java/com/hedera/services/cli/sign/RecordStreamSigningUtils.java

```diff
@@ -0,0 +1,287 @@
+import java.security.NoSuchProviderException;
+import java.security.SignatureException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * Utility class for signing record stream files
+ */
+public class RecordStreamSigningUtils {
+    /**
+     * Hidden constructor
+     */
+    private RecordStreamSigningUtils() {}
+
+    /**
+     * Supported stream version file
+     */
+    public static final int SUPPORTED_STREAM_FILE_VERSION = 6;
+
+    /**
+     * a messageDigest object for digesting entire stream file and generating entire record stream
+     * file hash
+     */
+    private static MessageDigest streamDigest;
+
+    /**
+     * a messageDigest object for digesting metaData in the stream file and generating metaData
+     * hash. Metadata contains: record stream version || HAPI proto version || startRunningHash ||
+     * endRunningHash || blockNumber, where || denotes concatenation
+     */
+    private static MessageDigest metadataStreamDigest;
+
+    /**
+     * Generates a signature file for the given record stream file
+     *
+     * @param signatureFileDestination the full path where the signature file will be generated
+     * @param streamFileToSign         the stream file to be signed
+     * @param keyPair                  the keyPair used for signing
+     * @param hapiVersion              the hapi protobuf version
+     * @return true if the signature file was generated successfully, false otherwise
+     */
+    public static boolean signRecordStreamFile(
+            @NonNull final Path signatureFileDestination,
+            @NonNull final Path streamFileToSign,
+            @NonNull final KeyPair keyPair,
+            @NonNull final String hapiVersion) {
+
+        Objects.requireNonNull(signatureFileDestination, "signatureFileDestination must not be null");
+        Objects.requireNonNull(streamFileToSign, "streamFileToSign must not be null");
+        Objects.requireNonNull(keyPair, "keyPair must not be null");
+        Objects.requireNonNull(hapiVersion, "hapiVersion must not be null");
+
+        try {
+            final var recordFile = streamFileToSign.toFile();
+            final int version = getRecordStreamVersion(recordFile);
+            if (version != SUPPORTED_STREAM_FILE_VERSION) {
+                System.err.printf(
+                        "signRecordStreamFile :: Failed to sign file [%s] with unsupported version [%s]%n",
+                        streamFileToSign.getFileName(), version);
+                return false;
+            }
+
+            initRecordDigest();
+            int[] fileHeader = createFileHeader(hapiVersion);
+
+            outputStreamDigest(fileHeader, recordFile.getAbsolutePath());
+            SignatureFile.Builder signatureFile = createSignatureFile(keyPair);
+            generateSigRecordStreamFile(signatureFileDestination.toFile(), signatureFile);
```

Comment:
These three lines are where you pretty much use those `static` message digest fields above.  This would be a normal use of _instance fields_ if this was a real class.  Or better (and if you want `static` methods for some reason) then there should be a _pipeline_ where the pair of digests are produced and passed as input arg and return values from one method to the next.

Reply:
Commented about Static before. If it's an instance field, yes, you can pass it in as parameters.